docs/rfcs/0012-introduce-ui-store.md

```diff
@@ -0,0 +1,161 @@
+engineView.requestApiResult { result ->
+  // received on the main thread.
+  store.dispatch(UpdateResultAction(result))
+}
+
+// store thread
+fun reduce(state: State, action: Action) {
+  is UpdateResultAction -> {
+    // do things here.
+  }
+}
+
+// store thread
+Middleware {
+  override fun invoke(
+    context: MiddlewareContext<State, Action>,
+    next: (Action) -> Unit,
+    action: Action,
+  ) {
+    // perform side-effects that also happen on the store thread.
+  }
+}
+
+// main thread
+store.flowScoped { flow ->
+  flow.collect {
+    // perform work on the main thread.
+  }
+}
+```
+
+With the changes in this RFC, this switching of threads can be reduced (notable comments marked with 📝):
+
+```kotlin
+// engine thread
+engineView.requestApiResult { result ->
+  // received on the main thread.
+  store.dispatch(UpdateResultAction(result))
+}
+
+// 📝 main thread - now on the same thread, processed immediately.
+fun reduce(state: State, action: Action) {
+  is UpdateResultAction -> {
+    // do things here.
+  }
+}
+
+// 📝 main thread - now on the same thread, processed immediately.
+Middleware {
+  override fun invoke(
+    context: MiddlewareContext<State, Action>,
+    next: (Action) -> Unit,
+    action: Action,
+  ) {
+    // 📝 perform side-effects that now happen on the main thread.
+  }
+}
+
+// main thread
+store.flowScoped { flow ->
+  flow.collect {
+      // perform work on the main thread.
+    }
+  }
+}
+```
+
+Additionally, from [performance investigations already done][2], we know that Fenix creates over a hundred threads within a few seconds of startup. Reducing the number of threads for Stores that do not have a strong requirement to run on a separate thread will lower the applications memory footprint.
+
+## Guide-level explanation
+
+Extending the existing `Store` class to use the `Dispatchers.Main.immediate` will ensure that UI stores will stay on the same UI thread and have that work done immediately. Using a distinct class named `UiStore` also makes it clear to the developer that this is work that will be done on the UI thread and its implications will be made a bit more clear when it's used.
+
+```kotlin
+@MainThread
+open class UiStore<S : State, A : Action>(
+    initialState: S,
+    reducer: Reducer<S, A>,
+    middleware: List<Middleware<S, A>> = emptyList(),
+    threadNamePrefix: String? = null,
+) : Store<S, A>(
+    initialState,
+    reducer,
+    middleware,
+    threadNamePrefix,
+) {
+  final override fun getDispatcher(): CoroutineDispatcher {
+    return Dispatchers.Main.immediate
+  }
+}
+
+// In the current Store, we would add `getDispatcher`.
+open class Store<S : State, A : Action>(
+    initialState: S,
+    reducer: Reducer<S, A>,
+    middleware: List<Middleware<S, A>> = emptyList(),
+    threadNamePrefix: String? = null,
+) {
+  internal open fun getDispatcher(): CoroutineDispatcher {
+    return Executors.newSingleThreadExecutor(threadFactory).asCoroutineDispatcher()
+  }
+}
+```
```

Comment:
Another alternative to this would be an internal constructor in which we can specify the dispatcher. Personally I would prefer that to an inheritance solution

```kotlin
open class Store<S : State, A : Action> internal constructor(
    initialState: S,
    reducer: Reducer<S, A>,
    middleware: List<Middleware<S, A>> = emptyList(),
    threadNamePrefix: String? = null,
    specifiedDispatcher: CoroutineDispatcher? = null,
) {
    
    constructor(
        initialState: S,
        reducer: Reducer<S, A>,
        middleware: List<Middleware<S, A>> = emptyList(),
        threadNamePrefix: String? = null,
    ) : this(initialState, reducer, middleware, threadNamePrefix, null)
    
    private val threadFactory = StoreThreadFactory(threadNamePrefix)
    private val dispatcher = specifiedDispatcher ?: Executors.newSingleThreadExecutor(threadFactory).asCoroutineDispatcher()
    private val reducerChainBuilder = ReducerChainBuilder(threadFactory, reducer, middleware)
    private val scope = CoroutineScope(dispatcher)

open class UiStore<S : State, A : Action>(
    initialState: S,
    reducer: Reducer<S, A>,
    middleware: List<Middleware<S, A>> = emptyList(),
    threadNamePrefix: String? = null,
) : Store<S, A>(
    initialState,
    reducer,
    middleware,
    threadNamePrefix,
    specifiedDispatcher = Dispatchers.Main.immediate
) {

```

Reply:
Thanks for documenting our offline discussion here, I forgot about this!

We didn't want to use this approach because it allows for overriding the dispatcher in the constructor which is the footgun we want to avoid anyone from using and why the RFC was updated not to include that as an option.

Reply:
Matt clarified offline, and I missed this (sorry!), that the default constructor is `internal` that has the dispatcher. So yes, this would work.

Both directions solve the same problem, so I have no preference between the styles. I'll update the RFC with this one, thanks!